state_synchronizer/src/peer_manager.rs

```diff
@@ -47,17 +48,29 @@ impl PeerManager {
             .into_iter()
             .map(|peer_id| (peer_id, PeerInfo::new(false, true, MAX_SCORE)))
             .collect();
+
         Self {
             peers,
             network_senders: HashMap::new(),
+            weighted_index: None,
         }
     }
 
     pub fn set_peers(&mut self, peer_ids: Vec<PeerId>) {
         let new_peer_ids: HashSet<_> = peer_ids.iter().collect();
+
+        let mut update_weighted_index = false;
         for (peer_id, info) in self.peers.iter_mut() {
-            info.is_upstream = new_peer_ids.contains(peer_id);
+            let new_is_upstream = new_peer_ids.contains(peer_id);
+            if info.is_upstream != new_is_upstream {
+                update_weighted_index = true;
+            }
+            info.is_upstream = new_is_upstream;
         }
+        if update_weighted_index {
```

Comment:
nit: IMO that's unnecessary optimization, cause set_peers should be relatively rare call (only used by Consensus) and building index is not a perf concern for it. I'd prefer readability here. Just  call compute_weighted_index unconditionally